File: source/ts/src/muhurta/engine.ts
import { getDailyPanchang } from '../core/panchang';
import { computeVaraTithiYogas } from './varaTithiYogas';
import { validateLocation, validateDate, validateLocalYearWindow } from '../utils/validation';
import { localYearWindow, civilDayStepper, clampToSupported } from '../utils/timezone';
import type { MuhurtaFactor } from './muhurtaTableTypes';
import type { GeoLocation } from '../types/location';
import type { Language, AyanamsaType, MasaSystem, PanchangSection } from '../types/options';
import type { DailyPanchangResult } from '../types/panchang';

/**
 * Rule indices: tithi 0..29 (0 = Shukla Pratipada, 29 = Amavasya), nakshatra 0..26
 * (Ashwini = 0), vara 0..6 (Sunday = 0), yoga 0..26 (Vishkambha = 0).
 */
export type { MuhurtaFactor } from './muhurtaTableTypes';

export interface MuhurtaRule {
  occasion: string;
  name?: string;
  auspiciousTithis?: readonly number[];
  inauspiciousTithis?: readonly number[];
  auspiciousNakshatras?: readonly number[];
  inauspiciousNakshatras?: readonly number[];
  auspiciousVaras?: readonly number[];
  inauspiciousVaras?: readonly number[];
  auspiciousYogas?: readonly number[];
  inauspiciousYogas?: readonly number[];
  /** `'ignore'` (the default), `'penalize'` (−15) or `'exclude'`, which vetoes seven whole tithis. */
  bhadra?: 'ignore' | 'penalize' | 'exclude';
  /** @deprecated Use {@link MuhurtaRule.bhadra}; when both are set, `bhadra` wins. */
  excludeBhadra?: boolean;
  /** Tithi 10 / 25 at sunrise. */
  excludeEkadashi?: boolean;
  requirePaksha?: 'shukla' | 'krishna';
  excludeAdhikaMasa?: boolean;
  excludeEclipse?: boolean;
  excludeGandaMula?: boolean;
  /** Only a Panchaka carrying a dosha vetoes; a Wednesday or Thursday spell ("Samanya") has none. */
  excludePanchaka?: boolean;
  /** Score the classical Vara x Tithi yogas at +10 / −15. Defaults to `true`. */
  varaTithiYogas?: boolean;
}

export interface MuhurtaScore {
  date: Date;
  /** 0..100. */
  score: number;
  /** Clears every hard exclusion and scores ≥ 50. */
  passes: boolean;
  /** English diagnostics: unlocalized, not a stable format, branch on `factors`. */
  reasons: string[];
  factors: MuhurtaFactor[];
}

export interface MuhurtaDay extends MuhurtaScore {
  panchang: DailyPanchangResult;
}

export interface MuhurtaScoreOptions {
  /** UTC offset in minutes, or an IANA zone name. */
  timezone: number | string;
  /** Defaults to `'lahiri'`. */
  ayanamsa?: AyanamsaType;
  language?: Language;
  /** Defaults to `'purnimanta'`. */
  masaSystem?: MasaSystem;
}

/** Everything {@link scoreFromPanchang} reads beyond the sections every day computes: the
 * eclipse, and the Bhadra that `lunarWindows` brings. No value depends on which sections were
 * asked for (INTERPOLATE_ALWAYS in `core/panchang.ts`), so these score a day as a full one does. */
const SCORED_SECTIONS: readonly PanchangSection[] = ['eclipse', 'lunarWindows'];

/**
 * Score one day against a muhurta rule; baseline 50, exclusions zero it outright.
 * @param date Any `Date` within the local calendar day to evaluate.
 */
export function scoreMuhurta(
  date: Date,
  location: GeoLocation,
  rule: MuhurtaRule,
  options: MuhurtaScoreOptions,
): MuhurtaScore {
  validateDate(date);
  validateLocation(location);
  const panchang = getDailyPanchang(date, location, {
    ...options,
    sections: SCORED_SECTIONS,
  });
  if (panchang === null) {
    return {
      date,
      score: 0,
      passes: false,
      reasons: ['polar location with no sunrise, Hindu day undefined'],
      factors: [{ code: 'no_sunrise', axis: 'exclusion', delta: 0 }],
    };
  }
  return scoreFromPanchang(panchang, rule);
}

/**
 * Every civil day of `options.timezone` from `start` to `end` inclusive, each at start's local
 * time of day, best score first; failures omitted unless asked for.
 */
export function computeAuspiciousDatesInRange(
  rule: MuhurtaRule,
  start: Date,
  end: Date,
  location: GeoLocation,
  options: MuhurtaScoreOptions & { includeFailures?: boolean },
): MuhurtaDay[] {
  validateDate(start);
  validateDate(end);
  validateLocation(location);
  if (start.getTime() > end.getTime()) {
    throw new RangeError(`start (${start.toISOString()}) must be ≤ end (${end.toISOString()})`);
  }
  return scoreCivilDays(rule, start.getTime(), end.getTime(), location, options);
}

export function scoreCivilDays(
  rule: MuhurtaRule,
  startMs: number,
  endMs: number,
  location: GeoLocation,
  options: MuhurtaScoreOptions & { includeFailures?: boolean },
): MuhurtaDay[] {
  const out: MuhurtaDay[] = [];
  const next = civilDayStepper(startMs, options.timezone);
  for (let t = next(); t <= endMs; t = next()) {
    const panchang = getDailyPanchang(new Date(clampToSupported(t)), location, options);
    if (panchang === null) continue;
    const result = scoreFromPanchang(panchang, rule);
    if (result.passes || options.includeFailures) {
      out.push({ ...result, panchang });
    }
  }

  out.sort((a, b) => b.score - a.score);
  return out;
}

/**
 * @internal {@link scoreCivilDays} for a caller that keeps only the scores, as `buildMuhurtaTable`
 * does: each day is built with {@link SCORED_SECTIONS} and no anga end times (the special yogas are
 * found from the tithi and nakshatra spans either way), which scores it exactly as the full day
 * does, and no panchang is kept. The same days in the same order.
 */
export function scoreOnlyCivilDays(
  rule: MuhurtaRule,
  startMs: number,
  endMs: number,
  location: GeoLocation,
  options: MuhurtaScoreOptions & { includeFailures?: boolean },
): MuhurtaScore[] {
  const out: MuhurtaScore[] = [];
  const dayOptions = { ...options, sections: SCORED_SECTIONS, computeEndTimes: false };
  const next = civilDayStepper(startMs, options.timezone);
  for (let t = next(); t <= endMs; t = next()) {
    const panchang = getDailyPanchang(new Date(clampToSupported(t)), location, dayOptions);
    if (panchang === null) continue;
    const result = scoreFromPanchang(panchang, rule);
    if (result.passes || options.includeFailures) out.push(result);
  }

  out.sort((a, b) => b.score - a.score);
  return out;
}

function scoreFromPanchang(p: DailyPanchangResult, rule: MuhurtaRule): MuhurtaScore {
  const reasons: string[] = [];
  const factors: MuhurtaFactor[] = [];
  let score = 50;
  let passes = true;

  const tithiAtSunrise = p.angas.tithis[0]!.index;
  const nakAtSunrise = p.angas.nakshatras[0]!.index;
  const yogaAtSunrise = p.angas.yogas[0]!.index;
  const varaIdx = p.angas.vara.index;

  const bhadraMode = rule.bhadra ?? (rule.excludeBhadra ? 'exclude' : 'ignore');

  if (bhadraMode === 'exclude' && p.inauspicious.bhadra !== null) {
    return zero(p.date, 'Bhadra Kala active on this day', 'bhadra');
  }
  if (rule.excludeEkadashi) {
    const isEkadashi = tithiAtSunrise === 10 || tithiAtSunrise === 25;
    if (isEkadashi) {
      return zero(p.date, 'Ekadashi tithi at sunrise', 'ekadashi');
    }
  }
  if (rule.excludeEclipse && p.eclipse !== null) {
    return zero(p.date, `Eclipse overlap (${p.eclipse.subtype})`, 'eclipse');
  }
  if (rule.excludeAdhikaMasa && p.calendar.chandramasa.isAdhika) {
    return zero(p.date, 'Adhika (intercalary) lunar month', 'adhika_masa');
  }
  if (rule.excludeGandaMula && p.inauspicious.gandaMula.active) {
    return zero(p.date, `Ganda Mula nakshatra (${p.inauspicious.gandaMula.severity})`, 'ganda_mula');
  }
  const pk = p.inauspicious.panchakaInfo;
  if (rule.excludePanchaka && pk.active && pk.isDosha) {
    return zero(p.date, `Panchaka active (${pk.type})`, 'panchaka');
  }
  if (rule.requirePaksha) {
    const paksha = tithiAtSunrise < 15 ? 'shukla' : 'krishna';
    if (paksha !== rule.requirePaksha) {
      return zero(p.date, `paksha is ${paksha}, rule requires ${rule.requirePaksha}`, 'paksha');
    }
  }

  if (rule.auspiciousTithis?.includes(tithiAtSunrise)) {
    score += 10;
    reasons.push(`auspicious tithi (${tithiAtSunrise})`);
    factors.push({ code: 'auspicious_tithi', axis: 'tithi', index: tithiAtSunrise, delta: 10 });
  } else if (rule.inauspiciousTithis?.includes(tithiAtSunrise)) {
    score -= 15;
    reasons.push(`inauspicious tithi (${tithiAtSunrise})`);
    factors.push({ code: 'inauspicious_tithi', axis: 'tithi', index: tithiAtSunrise, delta: -15 });
  }

  if (rule.auspiciousNakshatras?.includes(nakAtSunrise)) {
    score += 10;
    reasons.push(`auspicious nakshatra (${nakAtSunrise})`);
    factors.push({ code: 'auspicious_nakshatra', axis: 'nakshatra', index: nakAtSunrise, delta: 10 });
  } else if (rule.inauspiciousNakshatras?.includes(nakAtSunrise)) {
    score -= 15;
    reasons.push(`inauspicious nakshatra (${nakAtSunrise})`);
    factors.push({ code: 'inauspicious_nakshatra', axis: 'nakshatra', index: nakAtSunrise, delta: -15 });
  }

  if (rule.auspiciousVaras?.includes(varaIdx)) {
    score += 10;
    reasons.push(`auspicious vara (${varaIdx})`);
    factors.push({ code: 'auspicious_vara', axis: 'vara', index: varaIdx, delta: 10 });
  } else if (rule.inauspiciousVaras?.includes(varaIdx)) {
    score -= 15;
    reasons.push(`inauspicious vara (${varaIdx})`);
    factors.push({ code: 'inauspicious_vara', axis: 'vara', index: varaIdx, delta: -15 });
  }

  if (rule.auspiciousYogas?.includes(yogaAtSunrise)) {
    score += 10;
    reasons.push(`auspicious yoga (${yogaAtSunrise})`);
    factors.push({ code: 'auspicious_yoga', axis: 'yoga', index: yogaAtSunrise, delta: 10 });
  } else if (rule.inauspiciousYogas?.includes(yogaAtSunrise)) {
    score -= 15;
    reasons.push(`inauspicious yoga (${yogaAtSunrise})`);
    factors.push({ code: 'inauspicious_yoga', axis: 'yoga', index: yogaAtSunrise, delta: -15 });
  }

  if (rule.varaTithiYogas !== false) {
    for (const vty of computeVaraTithiYogas(varaIdx, tithiAtSunrise)) {
      const delta = vty.polarity === 'auspicious' ? 10 : -15;
      score += delta;
      reasons.push(`${vty.type} yoga (vara x tithi, ${vty.polarity})`);
      factors.push({ code: `vara_tithi_${vty.type}`, axis: 'varaTithiYoga', delta });
    }
  }

  if (bhadraMode === 'penalize' && p.inauspicious.bhadra !== null) {
    score -= 15;
    reasons.push('Bhadra Kala active during part of the day');
    factors.push({ code: 'bhadra', axis: 'karana', delta: -15 });
  }

  for (const sy of p.specialYogas) {
    if (sy.type === 'amrit_siddhi' || sy.type === 'sarvartha_siddhi'
      || sy.type === 'ravi_pushya' || sy.type === 'guru_pushya') {
      score += 5;
      reasons.push(`${sy.type} bonus`);
      factors.push({ code: sy.type, axis: 'specialYoga', delta: 5 });
    }
    if (sy.type === 'jwalamukhi') {
      score -= 10;
      reasons.push('Jwalamukhi yoga penalty');
      factors.push({ code: 'jwalamukhi', axis: 'specialYoga', delta: -10 });
    }
  }

  score = Math.max(0, Math.min(100, score));
  passes = score >= 50;

  return {
    date: p.date,
    score,
    passes,
    reasons,
    factors,
  };
}

function zero(date: Date, reason: string, code: string): MuhurtaScore {
  return {
    date,
    score: 0,
    passes: false,
    reasons: [reason],
    factors: [{ code, axis: 'exclusion', delta: 0 }],
  };
}

/** Every scored day in the local calendar year `year`, each queried at its local midnight. */
export function computeAuspiciousDatesForYear(
  year: number,
  rule: MuhurtaRule,
  location: GeoLocation,
  options: MuhurtaScoreOptions & { includeFailures?: boolean },
): MuhurtaDay[] {
  if (!Number.isInteger(year)) throw new RangeError(`year must be integer, got ${year}`);
  const [start, end] = localYearWindow(year, options.timezone);
  validateLocalYearWindow(year, start, end);
  validateLocation(location);
  return scoreCivilDays(rule, start, end, location, options);
}

/** @deprecated Renamed to {@link computeAuspiciousDatesInRange} in v5. */
export const findAuspiciousDates = computeAuspiciousDatesInRange;

File: source/go/internal/muhurta/engine.go
package muhurta

import (
	"context"
	"fmt"
	"sort"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/core"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

const dayMs int64 = 24 * 3600_000

var AllBhadraModes = []BhadraMode{BhadraIgnore, BhadraPenalize, BhadraExclude}

func ScoreMuhurta(
	eph *astronomy.EphemerisCtx,
	dateMs int64,
	location types.GeoLocation,
	rule MuhurtaRule,
	options MuhurtaScoreOptions,
) (MuhurtaScore, error) {
	if err := utils.ValidateDate(dateMs); err != nil {
		return MuhurtaScore{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return MuhurtaScore{}, err
	}
	opts := scoreOptionsPanchangOptions(options)
	opts.Sections = scorerSections()
	opts.SectionsGiven = true

	p, ok, err := core.GetDailyPanchang(eph, dateMs, location, opts, core.NatalResolvers{})
	if err != nil {
		return MuhurtaScore{}, err
	}
	if !ok {
		return MuhurtaScore{
			Date:    types.Date(dateMs),
			Score:   0,
			Passes:  false,
			Reasons: []string{"polar location with no sunrise, Hindu day undefined"},
			Factors: []MuhurtaFactor{{Code: "no_sunrise", Axis: AxisExclusion, Delta: 0}},
		}, nil
	}
	return scoreFromPanchang(p, rule)
}

func ComputeAuspiciousDatesInRange(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	rule MuhurtaRule,
	startMs, endMs int64,
	location types.GeoLocation,
	options MuhurtaScoreOptions,
) ([]MuhurtaDay, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(startMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(endMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	if startMs > endMs {
		return nil, types.Codef(types.ErrInvalidInput,
			"start (%s) must be ≤ end (%s)",
			types.Date(startMs).ISOString(), types.Date(endMs).ISOString())
	}
	return ScoreCivilDays(ctx, eph, rule, startMs, endMs, location, options)
}

// scorerSections are the optional sections scoreFromPanchang reads: the
// eclipse, and the lunar windows that carry Bhadra. The angas at sunrise, the
// vara, the lunar month, Ganda Mula, Panchaka and the special yogas it also
// reads are computed whatever the sections.
func scorerSections() core.SectionSet {
	return core.Sections(core.SectionEclipse, core.SectionLunarWindows)
}

// ScoreCivilDays scores one day per civil date of options.Timezone from
// startMs to endMs, each at startMs's local time of day, best score first.
func ScoreCivilDays(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	rule MuhurtaRule,
	startMs, endMs int64,
	location types.GeoLocation,
	options MuhurtaScoreOptions,
) ([]MuhurtaDay, error) {
	return scoreCivilDays(ctx, eph, rule, startMs, endMs, location, options,
		scoreOptionsPanchangOptions(options), true)
}

// scoreOnlyCivilDays is ScoreCivilDays for a caller that keeps only the
// scores, as BuildMuhurtaTable does: each day is built with the scorer's
// sections and no anga end times (the special yogas are found from the tithi
// and nakshatra spans either way), which scores it exactly as the full day
// does, and no MuhurtaDay keeps its Panchang.
func scoreOnlyCivilDays(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	rule MuhurtaRule,
	startMs, endMs int64,
	location types.GeoLocation,
	options MuhurtaScoreOptions,
) ([]MuhurtaDay, error) {
	opts := scoreOptionsPanchangOptions(options)
	opts.Sections = scorerSections()
	opts.SectionsGiven = true
	computeEndTimes := false
	opts.ComputeEndTimes = &computeEndTimes
	return scoreCivilDays(ctx, eph, rule, startMs, endMs, location, options, opts, false)
}

func scoreCivilDays(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	rule MuhurtaRule,
	startMs, endMs int64,
	location types.GeoLocation,
	options MuhurtaScoreOptions,
	opts types.PanchangOptions,
	keepPanchang bool,
) ([]MuhurtaDay, error) {
	out := []MuhurtaDay{}
	next, err := utils.CivilDayStepper(startMs, options.Timezone)
	if err != nil {
		return nil, err
	}
	for {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		t, err := next()
		if err != nil {
			return nil, err
		}
		if t > endMs {
			break
		}
		p, ok, err := core.GetDailyPanchang(eph, utils.ClampToSupported(t), location, opts, core.NatalResolvers{})
		if err != nil {
			return nil, err
		}
		if !ok {
			continue
		}
		result, err := scoreFromPanchang(p, rule)
		if err != nil {
			return nil, err
		}
		if result.Passes || options.IncludeFailures {
			day := MuhurtaDay{MuhurtaScore: result}
			if keepPanchang {
				day.Panchang = p
			}
			out = append(out, day)
		}
	}

	sort.SliceStable(out, func(i, j int) bool { return out[i].Score > out[j].Score })
	return out, nil
}

func scoreFromPanchang(p types.DailyPanchangResult, rule MuhurtaRule) (MuhurtaScore, error) {
	reasons := []string{}
	factors := []MuhurtaFactor{}
	score := 50

	if len(p.Angas.Tithis) == 0 || len(p.Angas.Nakshatras) == 0 || len(p.Angas.Yogas) == 0 {
		return MuhurtaScore{}, fmt.Errorf(
			"muhurta: panchang for %s has an empty anga list (tithis=%d nakshatras=%d yogas=%d)",
			p.Date.ISOString(), len(p.Angas.Tithis), len(p.Angas.Nakshatras), len(p.Angas.Yogas))
	}
	tithiAtSunrise := p.Angas.Tithis[0].Index
	nakAtSunrise := p.Angas.Nakshatras[0].Index
	yogaAtSunrise := p.Angas.Yogas[0].Index
	varaIdx := p.Angas.Vara.Index

	bhadraMode := ruleResolvedBhadra(rule)

	if bhadraMode == BhadraExclude && p.Inauspicious.Bhadra != nil {
		return zero(p.Date, "Bhadra Kala active on this day", "bhadra"), nil
	}
	if rule.ExcludeEkadashi {
		if tithiAtSunrise == 10 || tithiAtSunrise == 25 {
			return zero(p.Date, "Ekadashi tithi at sunrise", "ekadashi"), nil
		}
	}
	if rule.ExcludeEclipse && p.Eclipse != nil {
		return zero(p.Date,
			fmt.Sprintf("Eclipse overlap (%s)", p.Eclipse.Subtype), "eclipse"), nil
	}
	if rule.ExcludeAdhikaMasa && p.Calendar.Chandramasa.IsAdhika {
		return zero(p.Date, "Adhika (intercalary) lunar month", "adhika_masa"), nil
	}
	if rule.ExcludeGandaMula && p.Inauspicious.GandaMula.Active {
		return zero(p.Date,
			fmt.Sprintf("Ganda Mula nakshatra (%s)", p.Inauspicious.GandaMula.Severity),
			"ganda_mula"), nil
	}
	pk := p.Inauspicious.PanchakaInfo
	if rule.ExcludePanchaka && pk.Active && pk.IsDosha {
		return zero(p.Date, fmt.Sprintf("Panchaka active (%s)", pk.Type), "panchaka"), nil
	}
	if rule.RequirePaksha != "" {
		paksha := PakshaKrishna
		if tithiAtSunrise < 15 {
			paksha = PakshaShukla
		}
		if paksha != rule.RequirePaksha {
			return zero(p.Date,
				fmt.Sprintf("paksha is %s, rule requires %s", paksha, rule.RequirePaksha),
				"paksha"), nil
		}
	}

	type axis struct {
		auspicious, inauspicious []int
		index                    int
		name                     string
		factorAxis               MuhurtaFactorAxis
	}
	for _, a := range []axis{
		{rule.AuspiciousTithis, rule.InauspiciousTithis, tithiAtSunrise, "tithi", AxisTithi},
		{rule.AuspiciousNakshatras, rule.InauspiciousNakshatras, nakAtSunrise, "nakshatra", AxisNakshatra},
		{rule.AuspiciousVaras, rule.InauspiciousVaras, varaIdx, "vara", AxisVara},
		{rule.AuspiciousYogas, rule.InauspiciousYogas, yogaAtSunrise, "yoga", AxisYoga},
	} {
		switch {
		case containsInt(a.auspicious, a.index):
			score += 10
			reasons = append(reasons, fmt.Sprintf("auspicious %s (%d)", a.name, a.index))
			factors = append(factors, MuhurtaFactor{
				Code: "auspicious_" + a.name, Axis: a.factorAxis,
				Index: factorIndex(a.index), Delta: 10,
			})
		case containsInt(a.inauspicious, a.index):
			score -= 15
			reasons = append(reasons, fmt.Sprintf("inauspicious %s (%d)", a.name, a.index))
			factors = append(factors, MuhurtaFactor{
				Code: "inauspicious_" + a.name, Axis: a.factorAxis,
				Index: factorIndex(a.index), Delta: -15,
			})
		}
	}

	if ruleScoresVaraTithiYogas(rule) {
		yogas, err := ComputeVaraTithiYogas(varaIdx, tithiAtSunrise)
		if err != nil {
			return MuhurtaScore{}, err
		}
		for _, vty := range yogas {
			delta := -15
			if vty.Polarity == PolarityAuspicious {
				delta = 10
			}
			score += delta
			reasons = append(reasons,
				fmt.Sprintf("%s yoga (vara x tithi, %s)", vty.Type, vty.Polarity))
			factors = append(factors, MuhurtaFactor{
				Code: "vara_tithi_" + string(vty.Type), Axis: AxisVaraTithiYoga, Delta: delta,
			})
		}
	}

	if bhadraMode == BhadraPenalize && p.Inauspicious.Bhadra != nil {
		score -= 15
		reasons = append(reasons, "Bhadra Kala active during part of the day")
		factors = append(factors, MuhurtaFactor{Code: "bhadra", Axis: AxisKarana, Delta: -15})
	}

	for _, sy := range p.SpecialYogas {
		if sy.Type == types.YogaAmritSiddhi || sy.Type == types.YogaSarvarthaSiddhi ||
			sy.Type == types.YogaRaviPushya || sy.Type == types.YogaGuruPushya {
			score += 5
			reasons = append(reasons, string(sy.Type)+" bonus")
			factors = append(factors, MuhurtaFactor{
				Code: string(sy.Type), Axis: AxisSpecialYoga, Delta: 5,
			})
		}
		if sy.Type == types.YogaJwalamukhi {
			score -= 10
			reasons = append(reasons, "Jwalamukhi yoga penalty")
			factors = append(factors, MuhurtaFactor{
				Code: "jwalamukhi", Axis: AxisSpecialYoga, Delta: -10,
			})
		}
	}

	if score < 0 {
		score = 0
	}
	if score > 100 {
		score = 100
	}

	return MuhurtaScore{
		Date:    p.Date,
		Score:   score,
		Passes:  score >= 50,
		Reasons: reasons,
		Factors: factors,
	}, nil
}

func zero(date types.JSDate, reason, code string) MuhurtaScore {
	return MuhurtaScore{
		Date:    date,
		Score:   0,
		Passes:  false,
		Reasons: []string{reason},
		Factors: []MuhurtaFactor{{Code: code, Axis: AxisExclusion, Delta: 0}},
	}
}

func ComputeAuspiciousDatesForYear(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	year int,
	rule MuhurtaRule,
	location types.GeoLocation,
	options MuhurtaScoreOptions,
) ([]MuhurtaDay, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	startMs, endMs, err := utils.LocalYearWindow(year, options.Timezone)
	if err != nil {
		return nil, err
	}
	if err := utils.ValidateLocalYearWindow(year, startMs, endMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	return ScoreCivilDays(ctx, eph, rule, startMs, endMs, location, options)
}

func FindAuspiciousDates(
	ctx context.Context,
	eph *astronomy.EphemerisCtx,
	rule MuhurtaRule,
	startMs, endMs int64,
	location types.GeoLocation,
	options MuhurtaScoreOptions,
) ([]MuhurtaDay, error) {
	return ComputeAuspiciousDatesInRange(ctx, eph, rule, startMs, endMs, location, options)
}
